services/ansibler/server/domain/usecases/update_no_proxy_envs.go

```diff
@@ -0,0 +1,179 @@
+			return true
+		}
+	}
+
+	return false
+}
+
+func nodesChanged(currentK8sClusterInfo, desiredK8sClusterInfo *spec.ClusterInfo) bool {
+	currNodePool := currentK8sClusterInfo.GetNodePools()
+	desiredNodePool := desiredK8sClusterInfo.GetNodePools()
+
+	currStateNodesNum := 0
+	currStateIpMap := make(map[string]bool)
+	for _, np := range currNodePool {
+		for _, node := range np.Nodes {
+			currStateIpMap[node.Public] = true
+			currStateIpMap[node.Private] = true
+			currStateNodesNum += 1
+		}
+	}
+
+	desiredStateNodesNum := 0
+	for _, np := range desiredNodePool {
+		for _, node := range np.Nodes {
+			desiredStateNodesNum += 1
+			_, publicIPExists := currStateIpMap[node.Public]
+			_, privateIPExists := currStateIpMap[node.Private]
+
+			// Node in at least one nodepool was changed and no proxy env variables have to be changed.
+			if !publicIPExists || !privateIPExists {
+				return true
+			}
+		}
+	}
+
+	// Returns true if at lease one node was added or removed from the cluster.
+	// Otherwise returns false.
+	return desiredStateNodesNum != currStateNodesNum
+}
+
+// updateNoProxyEnvs handles the case where Claudie adds/removes node to/from the cluster.
+// Public and private IPs of this node must be added to the NO_PROXY and no_proxy env variables in
+// kube-proxy DaemonSet and static pods.
+func updateNoProxyEnvs(currentK8sClusterInfo, desiredK8sClusterInfo *spec.ClusterInfo, desiredLbs []*spec.LBcluster, spawnProcessLimit chan struct{}) error {
+	clusterID := commonUtils.GetClusterID(currentK8sClusterInfo)
+
+	// This is the directory where files (Ansible inventory files, SSH keys etc.) will be generated.
+	clusterDirectory := filepath.Join(baseDirectory, outputDirectory, fmt.Sprintf("%s-%s", clusterID, commonUtils.CreateHash(commonUtils.HashLength)))
+	if err := commonUtils.CreateDirectory(clusterDirectory); err != nil {
+		return fmt.Errorf("failed to create directory %s : %w", clusterDirectory, err)
+	}
+
+	if err := commonUtils.CreateKeysForDynamicNodePools(commonUtils.GetCommonDynamicNodePools(currentK8sClusterInfo.NodePools), clusterDirectory); err != nil {
+		return fmt.Errorf("failed to create key file(s) for dynamic nodepools : %w", err)
+	}
+
+	if err := commonUtils.CreateKeysForStaticNodepools(commonUtils.GetCommonStaticNodePools(currentK8sClusterInfo.NodePools), clusterDirectory); err != nil {
+		return fmt.Errorf("failed to create key file(s) for static nodes : %w", err)
+	}
+
+	noProxyList := createNoProxyList(desiredK8sClusterInfo.GetNodePools(), desiredLbs)
+	if err := utils.GenerateInventoryFile(templates.NoProxyEnvsInventoryTemplate, clusterDirectory, noProxyInventoryFileParameters{
+		K8sNodepools: NodePools{
+			Dynamic: commonUtils.GetCommonDynamicControlPlaneNodes(currentK8sClusterInfo.NodePools, desiredK8sClusterInfo.NodePools),
+			Static:  commonUtils.GetCommonStaticControlPlaneNodes(currentK8sClusterInfo.NodePools, desiredK8sClusterInfo.NodePools),
```

Comment:
Why do we need to extract common nodepools to both current and desired state ?
Why can't we use the nodepools from desired state ?

Reply:
We have to check if the control-plane nodes are in both desired and current state because if they aren't in the current state, they aren't in the k8s cluster yet, thus you can't run `kubeadm init` cmds to update no proxy envs on them. On the other hand, if they aren't in the desired state, they were already destroyed by the `terraformer`.